messaging/src/main/java/org/axonframework/eventhandling/async/PropertySequencingPolicy.java

```diff
@@ -0,0 +1,192 @@
+package org.axonframework.eventhandling.async;
+
+import static org.axonframework.common.BuilderUtils.assertNonNull;
+
+import java.util.function.Function;
+
+import org.axonframework.common.property.Property;
+import org.axonframework.common.property.PropertyAccessStrategy;
+import org.axonframework.eventhandling.EventMessage;
+
+/**
+ * A {@link SequencingPolicy} implementation that extracts the sequence identifier from the event message payload based
+ * on a given property or property extractor. If the event message payload is not of a supported type, a fallback 
+ * sequencing policy is used. By default the fallback sequencing policy raises an exception.
+ *
+ * @param <T>
+ *            The type of the supported event payloads.
+ * @param <K>
+ *            The type of the extracted property.
+ *
+ * @author Nils Christian Ehmke
+ * @since 4.5.2
+ */
+@SuppressWarnings( "rawtypes" )
+public final class PropertySequencingPolicy<T, K> implements SequencingPolicy<EventMessage> {
+
+	private final Class<T> payloadClass;
+	private final Function<T, K> propertyExtractor;
+	private final SequencingPolicy<EventMessage> fallbackSequencingPolicy;
+
+	@SuppressWarnings( "unchecked" )
+	private PropertySequencingPolicy( final Builder builder ) {
+		payloadClass = builder.payloadClass;
```

Comment:
I am missing a `Builder#validate` invocation. The infrastructure components do this to make sure all required fields are set.

Reply:
Ah, I spotted it in the `build()` method. The reason we place it in the constructor is in case inheritance is used on the component. Then we are no longer guaranteed the `build()` method is invoked since it may be overridden. The constructor would still be invoked though.

Reply:
Okay, I moved the validate invocation to the constructor.